File: go/sorting/bubble_sort.go
// Big O = O**2
package sorting

func BubbleSort(v []int) []int {
	for i := range v {
		// small optimization: iterate until unsorted position
		for j := 0; j < len(v)-i-1; j++ {
			if v[j] > v[j+1] {
				v[j], v[j+1] = v[j+1], v[j]
			}
		}
	}

	return v
}

File: rust/src/sorting/bubble_sort.rs
// Big O = O**2

pub fn bubble_sort(v: &mut Vec<i32>) -> &mut Vec<i32> {
    for i in 0..v.len() {
        // small optimization: iterate until unsorted position
        for j in 0..v.len() - i - 1 {
            if v[j] > v[j + 1] {
                v.swap(j, j + 1);
            }
        }
    }

    v
}
